torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -945,69 +945,12 @@ void ProcessGroupNCCL::enableCollectivesTiming() {
   enableTiming_.store(true);
 }
 
-std::future<bool> ProcessGroupNCCL::launchAsyncDebugDump() {
-  std::promise<bool> resultPromise;
-  std::future<bool> resultFuture = resultPromise.get_future();
-
-  std::thread workerThread(
-      [promise = std::move(resultPromise), this]() mutable {
-        try {
-          promise.set_value(dumpDebuggingInfo());
-        } catch (...) {
-          promise.set_exception(std::current_exception());
-        }
-      });
-
-  // Detach the thread to allow it to run independently
-  workerThread.detach();
-
-  return resultFuture;
-}
-
-std::chrono::time_point<std::chrono::steady_clock> getWakeupTime(
-    int intervalInMilSec) {
-  return std::chrono::steady_clock::now() +
-      std::chrono::milliseconds(intervalInMilSec);
-}
-
-void ProcessGroupNCCL::waitForDumpOrTimeout(
-    std::future<bool>& fut,
-    const std::chrono::time_point<std::chrono::steady_clock>& wakeUpTime,
-    size_t timeout_sec) {
-  TORCH_CHECK(fut.valid(), "Expected a valid future");
-
-  auto futStatus = fut.wait_for(std::chrono::seconds(timeout_sec));
-  TORCH_CHECK(
-      futStatus != std::future_status::deferred, "Expected eager launch.");
-  if (futStatus == std::future_status::ready) {
-    // Calling .get() will re-raise any exception from the future, and we don't
-    // care about the retval
-    try {
-      fut.get();
-      std::this_thread::sleep_until(wakeUpTime);
```

Comment:
Let me try to explain the reason for the old logic and then see if the new logic is adequate.

Timeout - we wait up this long for a dump before abandoning. Ideally we never abandon and the future completed well before the timeout expires 

Sleep- even in cases the local dump was fast, we wait a minimum time before letting our process exit such that mast won't kill other processes that may be still dumping.

Controversy- if we set the sleep too long it means job-restart time is impacted. We want jobs to restart asap to improve goodput.  But if we set sleep too short we may observe lost dump data from ranks killed by mast, and we want to never lose dump data.

Reply:
Thanks for explaining the original logic here!
Then there seems to be some duplications of logic here. E.g., In monitor thread, there is another 'sleep' serving the same purpose of waiting for the rank/or other ranks to dump `// Leave another two mins for desync report generation or process group
    // destroy.
    std::this_thread::sleep_for(std::chrono::seconds(heartbeatTimeoutInSec_));` Feel its better to remove this sleep logic from the general function of 'waitForFutureOrTimeout'. And leave this 'sleep' logic only in the last thread to join (monitoring thread).

Reply:
Ok that sounds good. I forgot if we still have a separate code path to dump from watchdog thread when collective timeout happens? Then it needs to sleep too.